packages/cozy-realtime/src/index.js

```diff
@@ -99,228 +120,233 @@ const configTypes = {
-        ])
-      },
-      unsubscribe: (doctype, event, listener) => {
-        if (
-          listeners[doctype] &&
-          listeners[doctype][event] &&
-          listeners[doctype][event].includes(listener)
-        ) {
-          listeners[doctype][event] = listeners[doctype][event].filter(
-            l => l !== listener
-          )
-        }
+      if (numRetries) {
+        console.warn(`Reconnecting ... ${numRetries} tries left.`)
+        setTimeout(() => {
+          try {
+            socket = connectWebSocket(
+              config,
+              onSocketMessage,
+              onSocketClose,
+              --numRetries,
+              retryDelay + 1000,
+              true
+            )
+          } catch (error) {
+            console.error(
+              `Unable to reconnect to realtime. Error: ${error.message}`
+            )
+          }
+        }, retryDelay)
       }
     }
+  }
 
-    resolve(cozySocket)
-  })
-}
+  try {
+    socket = connectWebSocket(
+      config,
+      onSocketMessage,
+      onSocketClose,
+      NUM_RETRIES,
+      RETRY_BASE_DELAY
+    )
+  } catch (error) {
+    throw error
+  }
+
+  return {
+    subscribe: (doctype, event, listener, docId) => {
+      if (typeof listener !== 'function')
+        throw new Error('Realtime event listener must be a function')
 
-// Returns the Promise of a subscription to a given doctype and document
-export function subscribe(config, doctype, doc, parse = doc => doc) {
-  return subscribeAll(config, doctype, parse).then(subscription => {
-    // We will call the listener only for the given document, so let's curry it
-    const docListenerCurried = listener => {
-      return syncedDoc => {
-        if (syncedDoc._id === doc._id) {
-          listener(syncedDoc)
-        }
+      const listenerKey = getListenerKey(doctype, docId)
+
+      if (!listeners.has(listenerKey)) {
+        listeners.set(listenerKey, {})
+        subscribeWhenReady(doctype, socket, docId)
       }
-    }
 
-    return {
-      onUpdate: listener => subscription.onUpdate(docListenerCurried(listener)),
-      onDelete: listener => subscription.onDelete(docListenerCurried(listener)),
-      unsubscribe: () => subscription.unsubscribe()
+      listeners.set(listenerKey, {
+        ...listeners.get(listenerKey),
+        [event]: [listener]
+      })
+
+      if (!subscriptionsState.has(listenerKey)) {
+        subscriptionsState.add(listenerKey)
+      }
+    },
+    unsubscribe: (doctype, event, listener, docId) => {
+      const listenerKey = getListenerKey(doctype, docId)
+      if (
+        listeners.has(listenerKey) &&
+        listeners.get(listenerKey)[event] &&
+        listeners.get(listenerKey)[event].includes(listener)
+      ) {
+        listeners.set(listenerKey, {
+          ...listeners.get(listenerKey),
+          [event]: listeners.get(listenerKey)[event].filter(l => l !== listener)
+        })
+      }
+      if (subscriptionsState.has(listenerKey)) {
```

Comment:
It looks wrong to me: if we had 2 subscriptions for the same doctype (or same doctype+docid), and we unsubcribe from the first, we should keep the `listenerKey` in `subscriptionsState` for the other subscription. I think we should check that `listeners.get(listenerKey)[event]` is empty before removing the key from `subscriptionsState`.

Reply:
Nope, it's more complicated than that: we should check that `listeners.get(listenerKey)` has empty arrays for all the possible events.